lib/web_ui/lib/src/engine/canvaskit/image_web_codecs.dart

```diff
@@ -0,0 +1,384 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/// Uses the `ImageDecoder` class supplied by the browser.
+///
+/// See also:
+///
+///  * `image_wasm_codecs.dart`, which uses codecs supplied by the CanvasKit WASM bundle.
+@JS()
+library image_web_codecs;
+
+import 'dart:async';
+import 'dart:js_util' as js_util;
+import 'dart:math' as math;
+import 'dart:typed_data';
+
+import 'package:js/js.dart';
+import 'package:ui/ui.dart' as ui;
+
+import 'canvaskit_api.dart';
+import 'image.dart';
+import 'surface.dart';
+import 'surface_factory.dart';
+
+@JS('window.ImageDecoder')
+external Object? get _imageDecoderConstructor;
+
+/// Whether the current browser supports `ImageDecoder`.
+bool browserSupportsImageDecoder =
+    _imageDecoderConstructor != null;
+
+/// Sets the value of [browserSupportsImageDecoder] to its default value.
+void debugResetBrowserSupportsImageDecoder() {
+  browserSupportsImageDecoder =
+      _imageDecoderConstructor != null;
+}
+
+/// Image decoder backed by the browser's `ImageDecoder`.
+class CkBrowserImageDecoder implements ui.Codec {
+  static Future<CkBrowserImageDecoder> create({
+    required Uint8List data,
+    required String debugSource,
+    int? targetWidth,
+    int? targetHeight,
+  }) async {
+    // ImageDecoder does not detect image type automatically. It requires us to
+    // tell it what the image type is.
+    final String? contentType = detectContentType(data);
+
+    if (contentType == null) {
+      throw ImageCodecException(
+        'Failed to detect image file format of the image data using the file header, which was:\n' +
+        data.sublist(0, math.min(10, data.length)).map((int byte) => byte.toRadixString(16).padLeft(2, '0')).join(' '),
+      );
+    }
+
+    if (!await _isImageTypeSupported(contentType)) {
+      throw ImageCodecException(
+        'Image file format $contentType is not supported by this browser\'s ImageDecoder.',
+      );
+    }
+
+    final _ImageDecoder webDecoder = _ImageDecoder(_ImageDecoderOptions(
+      type: contentType,
+      data: data,
+
+      // Flutter always uses premultiplied alpha.
+      premultiplyAlpha: 'premultiply',
+      desiredWidth: targetWidth,
+      desiredHeight: targetHeight,
+
+      // "default" gives the browser the liberty to convert to display-appropriate
+      // color space, typically SRGB, which is what we want.
+      colorSpaceConversion: 'default',
+
+      // Flutter doesn't give the developer to customize this, so if this is an
+      // animated image we should prefer the animated track.
+      preferAnimation: true,
+    ));
+
+    await js_util.promiseToFuture<void>(webDecoder.tracks.ready);
```

Comment:
This is all that's needed unless you want to ensure frame count is fully known.

Reply:
For some reason, until this morning (when I was forced to update my Chrome), if I didn't also `await ImageDecoder.completed` I would get different values for `VideoFrame.format`. It would flip between BGRX and I420. I assumed there was some race between some internal work that `ImageDecoder` does and the call to `decode()` (e.g. progressive loading of images).

With the latest version of Chrome I cannot reproduce this any more, but I'm not sure what behavior to expect. Since I changed the code to load the `VideoFrame` straight to `texImage2D`, it probably doesn't matter. We should be able to handle both. The only thing that might be surprising to the user (and perhaps to our goldens test infrastructure) is if the image is sometimes decoded at full resolution and sometimes partially.

Reply:
Ah, Chrome will opportunistically use YUV decoding, but it depends on if all the frame data is available. So waiting for completed will ensure it's YUV if the image format allows it and Chrome thinks it can decode to it.

In cases where a YUV capable image is decoded to BGRX it's still full resolution, it shouldn't muddle any pixel tests, but I'm not sure off the top of my head.